File: Python/55.Jump Game.py
#从第一个数出发，跳到最后一个数，当前位置的大小为可以跳的最大步数
import copy
class Solution(object):
    def canJump(self, nums):
        l , maxstep = len(nums) , 0
        for i in range(l):
            maxstep -= 1
            maxstep = max(maxstep,nums[i])
            if i+maxstep>=l-1:
                return True
            elif maxstep==0:
                return False

s = Solution()
print(s.canJump([5,9,3,2,1,0,2,3,3,1,0,0]))
print(s.canJump([1,0,1,0]))
print(s.canJump([0,2,3]))
print(s.canJump([2,0,0]))
print(s.canJump([2,3,1,1,4]))
print(s.canJump([3,2,1,0,4]))

File: C++/Array/55.Jump Game.cpp
/*
数组元素代表从该位置向后最多走的步数，求是否能到达最后一个。
*/

#include <iostream>
#include <vector>
#include <windows.h>
#include <windef.h>
using namespace std;

class Solution {
public:
	bool canJump(vector<int>& nums) {
		int length = nums.size();
		if(length<=1)	return true;
        for(int i=1; i < length-1; ++i)
        {
        	if(nums[i-1]==0)	return false;
        	nums[i] = max(nums[i-1]-1,nums[i]);
        }
        return nums[length-2]>0;
    }
};

int main()
{
	Solution s;
	//vector<int> nums{ 2,3,1,1,4};
	vector<int> nums{ 3,2,1,0,4};
	cout<<s.canJump(nums)<<endl;
}

